File: 2025/JuyoungMoon/week6/2025-07-13-maximum-depth-of-binary-tree.cpp
// Author: Juyoung Moon

// KDD LeetCode Study Week 6: Trees (DFS/BFS)
// https://github.com/juyomo/leetcode-study

// LeetCode #104.
// https://leetcode.com/problems/maximum-depth-of-binary-tree/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    if (root === null) {
        return 0;
    }
    return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
};

File: 2025/shinyoung/2025-07-13-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxDepth(self, root) -> int:
        if root is None:
            return 0

        def dfs(curr, count):
            if curr is None:
                return count

            if not curr.left and not curr.right:
                return count

            count += 1

            return max(dfs(curr.left, count), dfs(curr.right, count))

        return dfs(root, 1)


solution=Solution()
print(solution.maxDepth(TreeNode(0)))
print(solution.maxDepth(TreeNode(3, TreeNode(9, None, None), TreeNode(20, TreeNode(15), TreeNode(7)))))
